pkg/ruler/manager.go

```diff
@@ -334,15 +334,21 @@ func (r *DefaultMultiTenantManager) getOrCreateNotifier(userID string) (*notifie
 // the input shouldRemove() function returns true.
 func (r *DefaultMultiTenantManager) removeUsersIf(shouldRemove func(userID string) bool) {
 	r.userManagerMtx.Lock()
-	defer r.userManagerMtx.Unlock()
+
+	var wg sync.WaitGroup
 
 	// Check for deleted users and remove them
 	for userID, mngr := range r.userManagers {
 		if !shouldRemove(userID) {
 			continue
 		}
 
-		go mngr.Stop()
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			mngr.Stop()
```

Comment:
This could take a while as well, as this waits for inflight evaluations to finish, and so could also block further resharding.